### ISSUE TITLE
Support for weights as an optional parameter for did2s?

### ISSUE THREAD

**Issue by AnonTendim** [py-econometrics/pyfixest#557]
Similar to what is in R fixest, will you consider adding support for weights: Optional variable to run a weighted first- and second-stage regressions in did2s? Thanks!

**Comment by s3alfisc** [comment]
@rafimikail would this be something of interest for you to pick up?

**Comment by rafimikail** [comment]
Hey @s3alfisc , Sure, i will take on this task!

**Comment by s3alfisc** [comment]
Kyle's R journal paper is a great starting point to get familiar with the method: [link](https://journal.r-project.org/articles/RJ-2022-048/). 

Btw, thanks for the suggestion, @AnonTendim, very much appreciated!

**Comment by rafimikail** [comment]
I will check that one out, thanks for giving the reference :D

**Comment by marcandre259** [comment]
Apparently, there is a difference between pyfixest and fixest (R) when using the predict method with the new data argument. 

I'm still digging in this issue, but the reason behind it might be obvious. I'm getting different point estimates when including weights in pyfixest did2s and I figure this might be one of the reason. 

It seems like feols then gives the predictions of the unweighted regression. For example:

`
df_castle = pd.read_stata("https://github.com/scunning1975/mixtape/raw/master/castle.dta")
df_castle = df_castle.astype({"post": bool})

fit_untrt = pf.feols("l_homicide ~ 1 | state + year", data=df_untreated)
fit_untrt_wgt = pf.feols("l_homicide ~ 1 | state + year", data=df_untreated, weights="popwt")

yhat = fit_untrt.predict(newdata=df_castle)
yhat_wgt = fit_untrt_wgt.predict(newdata=df_castle)

# Getting first stage predictions
assert np.all(np.isclose(yhat[:5], yhat_wgt[:5]))

yhat = fit_untrt.predict()
yhat_wgt = fit_untrt_wgt.predict()

assert np.logical_not(np.all((np.isclose(yhat, yhat_wgt))))
`

**Comment by s3alfisc** [comment]
Hi @marcandre259 , thanks for raising this! 

If I understand you correctly, we would expect that with WLS, `predict()` would **always** produce different values than OLS, unless all weights are identical. .

The behavior of the predict() method for `newdata = None` is controlled here: https://github.com/py-econometrics/pyfixest/blob/fd8e1436efd14cc98071eed51b91f98624cd1b6c/pyfixest/estimation/feols_.py#L1606. 

If no new data is provided, `predict()` will return `self._Y_hat_link`, which is computed as the difference of the untransformed and unweighted Y via $Y - \hat{u} := \hat{Y}$:

 https://github.com/py-econometrics/pyfixest/blob/fd8e1436efd14cc98071eed51b91f98624cd1b6c/pyfixest/estimation/feols_.py#L457 

This seems correct to me. 

I suppose that the error with `newdata` is related to the use of fixed effects - if we drop them, we indeed end of with different results: 

```python
fit_untrt = pf.feols("l_homicide ~ 1 +C(state) + C(year)", data=df_untreated)
fit_untrt_wgt = pf.feols("l_homicide ~ 1 + C(state) + C(year)", data=df_untreated, weights="popwt")

yhat = fit_untrt.predict(newdata=df_castle)
yhat_wgt = fit_untrt_wgt.predict(newdata=df_castle)

yhat[0:5]
# array([1.96511559, 1.98852372, 1.96735659, 2.01274515, 2.00770564])
yhat_wgt [0:5]
# array([1.98343013, 2.0046152 , 1.99374426, 2.01362497, 1.98860927])
```
So the error must lie somewhere here https://github.com/py-econometrics/pyfixest/blob/fd8e1436efd14cc98071eed51b91f98624cd1b6c/pyfixest/estimation/feols_.py#L1614. 

I will take a look at this now. Thanks for raising this!

**Comment by s3alfisc** [comment]
Hm, I am actually a bit confused now - running under pyfixest 0.25.2 (the current PyPi release), PyFixest always matches fixest for "empty" models: 

## no weights, no newdata
```{python}
import pyfixest as pf
import pandas as pd
pd.options.display.precision = 7

data = pf.get_data()

vcov = "hetero"

fit = pf.feols(
  fml = "Y ~ 1 | f1 + f2",
  data = data, 
)

fit.predict()[0:5]
# array([-1.59872466,  3.10764036, -0.45737367, -0.18200548, -1.92886362])
```

```{r}
library(fixest)
library(reticulate)

fit = feols(
  fml = Y ~ 1 | f1 + f2,
  data = py$data, 
)

predict(fit)[0:5]
# [1] -1.5987247  3.1076404 -0.4573737 -0.1820055 -1.9288636
```

## weights, no newdata
```{python}
fit = pf.feols(
  fml = "Y ~ 1 | f1 + f2",
  data = data, 
  weights = "weights"
)

fit.predict()[0:5]
#array([-1.4485639 ,  3.06560124, -0.35331312, -0.16466192, -2.14148705])
```

```{r}
fit = feols(
  fml = Y ~ 1 | f1 + f2,
  data = py$data, 
  weights = ~weights
)

predict(fit)[0:5]
# [1] -1.4485639  3.0656012 -0.3533131 -0.1646619 -2.1414870
```
## no weights, newdata
```{python}
fit = pf.feols(
  fml = "Y ~ 1 | f1 + f2",
  data = data.iloc[0:100], 
)

fit.predict()[0:5]
# array([-1.67082186,  3.75392669,  0.13442001, -0.26977319, -1.54265443])
```

```{r}
fit = feols(
  fml = Y ~ 1 | f1 + f2,
  data = py$data[1:100,], 
)

predict(fit)[0:5]
# [1] -1.6708219  3.7539267  0.1344200 -0.2697732 -1.5426544

```
## weights, newdata
```{python}
fit = pf.feols(
  fml = "Y ~ 1 | f1 + f2",
  data = data.iloc[0:100], 
  weights = "weights"
)

fit.predict()[0:5]
# array([-1.36798072,  3.09052573,  0.13442001, -0.32484456, -1.82236608])
```

```{r}
fit = feols(
  fml = Y ~ 1 | f1 + f2,
  data = py$data[1:100,], 
  weights = ~weights
)

predict(fit)[0:5]
# [1] -1.3679807  3.0905250  0.1344200 -0.3248445 -1.8223663
```
Which version are you running @marcandre259 ?

**Comment by s3alfisc** [comment]
This is a bug in WLS with newdata provided, and I have opened an issue here: #678

**Comment by s3alfisc** [comment]
Can be closed as we merged #692 thanks to @marcandre259's work =)
